core/ecschema-editing/src/Editing/RelationshipClasses.ts

```diff
@@ -117,6 +117,50 @@ export class RelationshipClasses extends ECClasses {
     return { itemKey: newClass.key };
   }
 
+  /**
+   * Sets the base class of a RelationshipClass.
+   * @param relationshipKey The SchemaItemKey of the RelationshipClass.
+   * @param baseClassKey The SchemaItemKey of the base class. Specifying 'undefined' removes the base class.
+   */
+  public async setBaseClass(relationshipKey: SchemaItemKey, baseClassKey?: SchemaItemKey): Promise<SchemaItemEditResults> {
+    const relClass = await this._schemaEditor.schemaContext.getSchemaItem<MutableRelationshipClass>(relationshipKey);
+    if (relClass === undefined)
+      return { itemKey: relationshipKey, errorMessage: `Relationship Class ${relationshipKey.fullName} not found in schema context.` };
+
+    if (baseClassKey === undefined) {
+      relClass.baseClass = undefined;
+      return { itemKey: relationshipKey };
+    }
+
+    const baseClassSchema = baseClassKey.schemaKey.matches(relationshipKey.schemaKey)
+      ? relClass.schema
+      : await this._schemaEditor.getSchema(baseClassKey.schemaKey);
+
+    if (baseClassSchema === undefined) {
+      return { itemKey: relationshipKey, errorMessage: `Schema Key ${baseClassKey.schemaKey.toString(true)} not found in context` };
+    }
+
+    const baseClass = await baseClassSchema.lookupItem<RelationshipClass>(baseClassKey);
+    if (baseClass === undefined)
+      return { itemKey: relationshipKey, errorMessage: `Unable to locate base class ${baseClassKey.fullName} in schema ${baseClassSchema.fullName}.` };
+
+    if (baseClass.schemaItemType !== SchemaItemType.RelationshipClass)
+      return { itemKey: relationshipKey, errorMessage: `${baseClass.fullName} is not of type Relationship Class.` };
+
+    if (relClass.baseClass !== undefined && !await baseClass.is(await relClass.baseClass))
+      return { itemKey: relationshipKey, errorMessage: `Baseclass ${baseClass.fullName} must derive from ${relClass.baseClass.fullName}.`};
+
+    const currentBaseClass = relClass.baseClass;
+    relClass.baseClass = new DelayedPromiseWithProps<SchemaItemKey, RelationshipClass>(baseClassKey, async () => baseClass);
+    const result = await this.validate(relClass);
```

Comment:
Can everything above this line be moved to a setBaseClass method in ECClasses, allowing the same code for setBaseClass for entities, mixins, relationships, etc...?  You can then override it if necessary, like the needed validation for relationships.  Call the base method, perform validation in the override, and the reset the base class in case of a rule violation.